// vscode/vscode/src/vs/workbench/api/common/extHostSpeech.js
import { CancellationTokenSource } from 'monaco-editor/esm/vs/base/common/cancellation.js';
import { toDisposable, DisposableStore } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
import { MainContext } from './extHost.protocol.js';
class ExtHostSpeech {
    static { this.ID_POOL = 1; }
    constructor(mainContext) {
        this.providers = ( new Map());
        this.sessions = ( new Map());
        this.proxy = ( mainContext.getProxy(MainContext.MainThreadSpeech));
    }
    async $createSpeechToTextSession(handle, session) {
        const provider = this.providers.get(handle);
        if (!provider) {
            return;
        }
        const disposables = ( new DisposableStore());
        const cts = ( new CancellationTokenSource());
        this.sessions.set(session, cts);
        const speechToTextSession = disposables.add(provider.provideSpeechToTextSession(cts.token));
        disposables.add(speechToTextSession.onDidChange(e => {
            if (cts.token.isCancellationRequested) {
                return;
            }
            this.proxy.$emitSpeechToTextEvent(session, e);
        }));
        disposables.add(cts.token.onCancellationRequested(() => disposables.dispose()));
    }
    async $cancelSpeechToTextSession(session) {
        this.sessions.get(session)?.dispose(true);
        this.sessions.delete(session);
    }
    registerProvider(extension, identifier, provider) {
        const handle = ExtHostSpeech.ID_POOL++;
        this.providers.set(handle, provider);
        this.proxy.$registerProvider(handle, identifier, { extension, displayName: extension.value });
        return toDisposable(() => {
            this.proxy.$unregisterProvider(handle);
            this.providers.delete(handle);
        });
    }
}
export { ExtHostSpeech };
